### PR TITLE
feat(std/http): Check if cookie property is valid

### PR DESCRIPTION
**Pull request by getspooky**

### CODE REVIEW

**Comment by ry** [comment]
Looks good - but can you add a test?

File: std/http/cookie_test.ts [hunk 1]
```diff
@@ -31,6 +31,28 @@ Deno.test({
...
+        name: "Name",
+        value: "中文名字",
+        httpOnly: true,
+        secure: true,
+        maxAge: 0,
+      });
+    } catch (e) {
+      error = true;
+    }
+    assert(error);
```
Comment (schwarzkopfb) [inline std/http/cookie_test.ts:51]: [`assertThrows()`](https://github.com/denoland/deno/blob/master/std/testing/asserts.ts#L425) from `std/testing/asserts` is a powerful tool for these kind of checks. 😉

Comment (getspooky) [reply std/http/cookie_test.ts:51]: Hi @schwarzkopfb 
If you like all testing file used in deno. They use this approach.

Comment (nayeemrmn) [reply std/http/cookie_test.ts:51]: @getspooky `assertThrows()` is the way to go.

Comment (getspooky) [reply std/http/cookie_test.ts]: @nayeemrmn when i run deno test std/http/cookie_test.ts i got 
```sh
TS2345 [ERROR]: Argument of type '{ depth: number; sorted: boolean; trailingComma: boolean; compact: boolean; iterableLimit: number; }' is not assignable to parameter of type 'InspectOptions'.
  Object literal may only specify known properties, and 'sorted' does not exist in type 'InspectOptions'.
      sorted: true,
      ~~~~~~~~~~~~
    at file:///home/getspooky/Documents/OpenSourceProject/deno/std/testing/asserts.ts:26:7
```

Comment (bartlomieju) [reply std/http/cookie_test.ts]: @getspooky looks like you're running on old version of Deno

Comment (getspooky) [reply std/http/cookie_test.ts]: @bartlomieju thanks , I forgot to update deno version

**Comment by getspooky** [comment]
@ry I added test for cookie propery validation.

**Comment by bartlomieju** [comment]
@getspooky CI is failing, try running again by pushing empty commit: `git commit -m "reset CI" --allow-empty`

**Comment by getspooky** [comment]
@ry I added test for cookie validation

**Comment by bartlomieju** [comment]
@nayeemrmn PTAL

File: std/http/cookie.ts [hunk 2]
```diff
@@ -32,11 +32,14 @@ export interface Cookie {
 
 export type SameSite = "Strict" | "Lax" | "None";
 
+const FIELD_CONTENT_REGEXP = /^[\u0009\u0020-\u007e\u0080-\u00ff]+$/;
```
Comment (nayeemrmn) [inline std/http/cookie.ts:35]: Where is this character set from?

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie gives valid character sets for cookie names and values, neither of which are this. I assume `domain` and `path` are encoded like they are in URLs.

Comment (getspooky) [reply std/http/cookie.ts:35]: @nayeemrmn what about other browser like [internet explorer]() and [Safari - Apple](https://www.apple.com/safari/) ?

Comment (nayeemrmn) [reply std/http/cookie.ts:35]: > @nayeemrmn what about other browser like internet explorer and [Safari - Apple](https://www.apple.com/safari/) ?

What about them? Is the MDN page I linked not consistent with the relevant spec?

Comment (getspooky) [reply std/http/cookie.ts:35]: @nayeemrmn So we should remove validation property from path and domain ?

Comment (lucacasonato) [reply std/http/cookie.ts:35]: @getspooky Could you use the character set specified on MDN (or the spec) for each of the values being checked? I don't know where you got this regexp from, but it doesn't look right.

Comment (getspooky) [reply std/http/cookie.ts:35]: @lucacasonato https://github.com/pillarjs/cookies/blob/master/index.js

Comment (getspooky) [reply std/http/cookie.ts:35]: @lucacasonato  any feedback ?

Comment (kitsonk) [reply std/http/cookie.ts:35]: From MDN:

> - A `<cookie-name>` can be any US-ASCII characters, except control characters, spaces, or tabs. It also must not contain a separator character like the following: `( ) < > @ , ; : \ " / [ ] ? = { }`.
> - A `<cookie-value>` can optionally be wrapped in double quotes and include any US-ASCII characters excluding control characters, Whitespace, double quotes, comma, semicolon, and backslash.

This RegEx isn't either...  It allows way too much.  There is no good in partially filtering the list.  So the feedback is we need something that follows the standards if we are going to do it.

**Review by bartlomieju — CHANGES_REQUESTED** [review]
https://github.com/denoland/deno/pull/7189#discussion_r510493616

File: std/http/cookie.ts (hunk 2 above)
Comment (getspooky) [reply std/http/cookie.ts:35]: RegExp to match field-content in RFC 7230 sec 3.2 
 * field-content = field-vchar [ 1*( SP / HTAB ) field-vchar ]
 * field-vchar   = VCHAR / obs-text
 * obs-text      = %x80-FF
```js
/^[\u0009\u0020-\u007e\u0080-\u00ff]+$/
```

**Comment by bartlomieju** [comment]
@nayeemrmn @lucacasonato I think we could land this PR now, any objections?

File: std/http/cookie.ts [hunk 3]
```diff
@@ -61,12 +64,14 @@ function toString(cookie: Cookie): string {
     out.push(`Max-Age=${cookie.maxAge}`);
   }
   if (cookie.domain) {
+    validateCookieProperty("Domain", cookie.domain);
```
Comment (nayeemrmn) [inline std/http/cookie.ts:67]: Remove.
```suggestion
```

File: std/http/cookie.ts [hunk 4]
```diff
@@ -61,12 +64,14 @@ function toString(cookie: Cookie): string {
...
     out.push(`Domain=${cookie.domain}`);
   }
   if (cookie.sameSite) {
     out.push(`SameSite=${cookie.sameSite}`);
   }
   if (cookie.path) {
+    validateCookieProperty("Path", cookie.path);
```
Comment (nayeemrmn) [inline std/http/cookie.ts:74]: Remove.
```suggestion
```

File: std/http/cookie.ts [hunk 5]
```diff
@@ -79,6 +84,20 @@ function toString(cookie: Cookie): string {
...
+ * @param value Value of the cookie.
+ */
+function validateCookieProperty(
+  key: string,
+  value: string | undefined | null,
+): void {
+  if (value && !FIELD_CONTENT_REGEXP.test(value)) {
+    throw new TypeError(`The ${key} of the cookie (${value}) is invalid.`);
+  }
+}
```
Comment (nayeemrmn) [inline std/http/cookie.ts:99]: This is only good for cookie names, rename.
```suggestion
function validateCookieName(value: string | undefined | null): void {
  if (value && !FIELD_CONTENT_REGEXP.test(value)) {
    throw new TypeError(`Invalid cookie name: "${value}".`);
  }
}
```

File: std/http/cookie.ts [hunk 6]
```diff
@@ -32,11 +32,14 @@ export interface Cookie {
...
 export type SameSite = "Strict" | "Lax" | "None";
 
+const FIELD_CONTENT_REGEXP = /^[^()@<>,;:\\"\[\]?={}\s]+$/;
+
 function toString(cookie: Cookie): string {
   if (!cookie.name) {
     return "";
   }
   const out: string[] = [];
+  validateCookieProperty("Name", cookie.name);
```
Comment (nayeemrmn) [inline std/http/cookie.ts:42]: Apply rename.
```suggestion
  validateCookieName(cookie.name);
```

File: std/http/cookie.ts [hunk 7]
```diff
@@ -32,11 +32,14 @@ export interface Cookie {
 
 export type SameSite = "Strict" | "Lax" | "None";
 
+const FIELD_CONTENT_REGEXP = /^[^()@<>,;:\\"\[\]?={}\s]+$/;
```
Comment (nayeemrmn) [inline std/http/cookie.ts:35]: Restrict to ASCII, excluding control characters.
```suggestion
const FIELD_CONTENT_REGEXP = /^(?=[\x20-\x7E]*$)[^()@<>,;:\\"\[\]?={}\s]+$/;
```

**Review by nayeemrmn — CHANGES_REQUESTED** [review]
One these are addressed we'll have a valid fix for cookie names only.

**Comment by getspooky** [comment]
@nayeemrmn I made the changes

**Review by nayeemrmn — APPROVED** [review]
LGTM

**Review by bartlomieju — COMMENTED** [review]
LGTM as well, but I'd be great to get a few more tests

File: std/http/cookie_test.ts [hunk 8]
```diff
@@ -31,6 +31,27 @@ Deno.test({
...
 
+Deno.test({
+  name: "Cookie Validation",
+  fn(): void {
+    const res: Response = {};
+    res.headers = new Headers();
+    assertThrows(
+      (): void => {
+        setCookie(res, {
+          name: "Space@",
```
[the remark this answers is not in the document]
Comment (getspooky) [reply std/http/cookie_test.ts]: @bartlomieju I added a list of tests that checks all cases.

**Review by bartlomieju — APPROVED** [review]
LGTM, thank you @getspooky

**Pull request merged**